packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+    if (this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlob to BlockInputBlobs after it already is complete"
+      );
+    }
+
+    // this check suffices for checking slot, parentRoot, and forkName
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source,
+          peerId: `${peerIdStr}`,
+        },
+        "Blob BeaconBlockHeader blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    if (this.state.hasBlock) {
+      assertBlockAndBlobArePaired(this.blockRootHex, this.state.block, blobSidecar);
+    }
+
+    // TODO: (@matthewkeil) check for duplicates and add metric here
+    // if (this.blobsCache.has(blobSidecar.index)) {
+    //   this.metrics.blockInput.duplicateBlob.inc()
+    // }
+
+    this.blobsCache.set(blobSidecar.index, {blobSidecar, source, seenTimestampSec, peerIdStr});
+
+    if (this.state.hasBlock && this.blobsCache.size === this.state.block.message.body.blobKzgCommitments.length) {
+      this.state = {
+        ...this.state,
+        hasAllData: true,
+        timeCompleteSec: seenTimestampSec,
+      };
+      this.dataPromise.resolve([...this.blobsCache.values()].map(({blobSidecar}) => blobSidecar));
+    }
+  }
+
+  getMissingBlobMeta(): BlobMeta[] {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INCOMPLETE_DATA,
+          ...this.getLogMeta(),
+        },
+        "Cannot get missing blobs. Block is unknown"
+      );
+    }
+    if (this.state.hasAllData) {
+      return [];
+    }
+
+    const blobMeta: BlobMeta[] = [];
+    const versionHashes = this.state.versionHashes;
+    for (let index = 0; index < versionHashes.length; index++) {
+      if (!this.blobsCache.has(index)) {
+        blobMeta.push({
+          index,
+          blockRoot: fromHex(this.blockRootHex),
+          versionHash: versionHashes[index],
+        });
+      }
+    }
+    return blobMeta;
+  }
+
+  getAllBlobsWithSource(): BlobWithSource[] {
+    if (!this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INCOMPLETE_DATA,
+          ...this.getLogMeta(),
+        },
+        "Cannot get all blobs. DA status is not complete"
+      );
+    }
+    return [...this.blobsCache.values()];
+  }
+
+  getBlobs(): deneb.BlobSidecars {
+    return this.getAllBlobsWithSource().map(({blobSidecar}) => blobSidecar);
+  }
+}
+
+function getVersionHashes(block: SignedBeaconBlock<ForkPostDeneb>): VersionedHashes {
+  return block.message.body.blobKzgCommitments.map(kzgCommitmentToVersionedHash);
```

Comment:
```suggestion
  return this.state.versionedHashes;
```

Reply:
this is a function to compute versioned hashes. I suppose its a oneliner and can be deleted and inlined